### PR TITLE
Add constant and percentage-based fee limits to payments

### PR DESCRIPTION
**Pull request by sebastiandelgado**
Issue #278

Status: WIP

I made this initial commit for this PR with a solution that has some assumptions (described below) that I'd like to discuss before settling on them and writing tests. That said, I think this isn't far off from the correct implementation of this feature. Let me know if this implementation significantly differs from what you had in mind.

I should also mention that some of the code in this PR is based on the PR issued by @djseeds last year (#320). I attempted to rebase his branch in order to properly give him credit but every single file had significant conflicts since it was last rebased mid-October. Please advise on an appropriate way to give him credit for his work.

# Complete:
 - Extend SendRequest and QueryRoutesRequest protos
 - newRoute function takes fee limit and cuts off routes that exceed it 
 - queryRoutes, payInvoice and sendPayment commands take the feeLimit inputs and pass them down to newRoute
 - When no feeLimit is included, don't enforce any feeLimits at all (by setting feeLimit to MaxInt64Value)

 
# Incomplete:
 - No tests yet
 - Haven't figured out how to pass down the FeeLimit arg from  sendPaymentRequest (commands.go) to rpcServer.SendPayment/SendPaymentSync (rpcServer.go). Still figuring out how/where to get the FeeLimit from the paymentStream input of rpcServer.SendPayment/SendPaymentSync. Any comments/pointers are very welcome here.

# Need input on:
 - Do we want to extend the functionality of queryRoutes and payInvoice to use the feeLimit as I did in this PR? It wasn't specified in the issue but seems like it would be useful.
 - If no feeLimit is specified, do we want to automatically set the fee limit to the payment amount (which is what I'm defaulting to in the test files) or should we just not enforce a limit by setting the limit to max(int64) (as is implemented now)?

### CODE REVIEW

**Comment by MeshCollider** [comment]
> Please advise on an appropriate way to give him credit for his work.

You can add him as a co-author, see https://github.com/blog/2496-commit-together-with-co-authors

File: cmd/lncli/commands.go [hunk 1]
```diff
@@ -1069,9 +1109,12 @@ func sendPayment(ctx *cli.Context) error {
...
 		}
 
+		feeLimit = calculateFeeLimit(ctx, amount)
+
 		req = &lnrpc.SendRequest{
-			Dest: destNode,
-			Amt:  amount,
+			Dest:     destNode,
+			Amt:      amount,
+			FeeLimit: feeLimit,
```
Comment (danrobinson) [inline cmd/lncli/commands.go]: I think the FeeLimit may need to be computed in the struct on line 1072 as well (in the case where a payment request is being fulfilled).

**Comment by danrobinson** [comment]
I think there may be a few additional changes that need to be made in `rpcserver.go`. Currently it loses the `feeLimit` field before it passes the payment on to `router.SendPayment`: https://github.com/lightningnetwork/lnd/pull/714/files#diff-01ada5680aaa18fc79cbf78b2d55a37bR1762

File: cmd/lncli/commands.go [hunk 2]
```diff
@@ -1178,9 +1231,14 @@ func payInvoice(ctx *cli.Context) error {
 		return fmt.Errorf("pay_req argument missing")
 	}
 
+	amt = ctx.Int64("amt")
```
Comment (danrobinson) [inline cmd/lncli/commands.go]: I think there's a problem here if the user passes a payment request ID, in which case they generally will not pass an `amt` argument (so this will be 0). For this reason, you may have to do the percentage-fee-limit calculation on the backend, in `rpcserver.go`.

File: cmd/lncli/commands.go [hunk 3]
```diff
@@ -84,6 +84,37 @@ func actionDecorator(f func(*cli.Context) error) func(*cli.Context) error {
 	}
 }
 
+// calculateFeeLimit takes ctx and the amount for the payment in satoshis.
+// It checks the fee_limit and fee_limit_percent flags and determines the
+// the feeLimit based on those values. If both flags are set, calculateFeeLimit
+// sets feeLimit to the Minumum value between them. If neither flag is set,
+// feeLimit is set to the maximum int64 value as to avoid enforcing a limit.
```
Comment (danrobinson) [inline cmd/lncli/commands.go]: Maybe there _should_ be a default limit, for safety? Paying fees that are a huge multiple of the amount transferred might be necessary and desirable in some extreme circumstances, but maybe the user should be protected from doing it unintentionally.

File: lnrpc/rpc.swagger.json [hunk 4]
```diff
@@ -696,6 +704,15 @@
...
   "definitions": {
+    "NewAddressRequestAddressType": {
+      "type": "string",
+      "enum": [
+        "WITNESS_PUBKEY_HASH",
+        "NESTED_PUBKEY_HASH",
+        "PUBKEY_HASH"
+      ],
+      "default": "WITNESS_PUBKEY_HASH"
+    },
```
Comment (danrobinson) [inline lnrpc/rpc.swagger.json]: Is this related to the rest of the pull request? It seems like it may be an unintentional reversion of an [old change](https://github.com/danrobinson/lnd/commit/129a0e10dec465d6f83c5bdd4464358172c1d6af#diff-b462ea9f4b73ab206027a14b3903d0cbL583).

File: routing/router.go [hunk 5]
```diff
@@ -1361,6 +1362,9 @@ type LightningPayment struct {
 	// used.
 	FinalCLTVDelta *uint16
 
+	// FeeLimit is a user-specified maximum fee for this payment in satoshis.
+	FeeLimit btcutil.Amount
```
Comment (danrobinson) [inline routing/router.go]: Why not express FeeLimit in millisatoshis throughout, to match `amt`?

File: lnrpc/rpc.swagger.json (hunk 4 above)
Comment (sebastiandelgado) [reply lnrpc/rpc.swagger.json]: Yep, you're right about this. lnrpc/gen_protos.sh generates a bunch of code that corresponds to older commits and I missed getting rid of this one. Changing this.

File: cmd/lncli/commands.go (hunk 3 above)
Comment (sebastiandelgado) [reply cmd/lncli/commands.go]: Ok, should this limit be an absolute number or proportional to transaction amount? 

Off the top of my head, I'm thinking limit should be equal to, or some small multiple (2x or 3x) of the transaction amount. I say this because although there are circumstances in which someone might want to pay fees equal to or larger than the transaction amount, it seems reasonable to make users explicitly declare they want to go forward with the transaction.

Open to implementing this according to whichever rules you guys suggest though.

**Comment by sebastiandelgado** [comment]
@danrobinson thanks for the thoughtful comments! Will be addressing them over the weekend.

I noted changes you mentioned in rpcserver.go as incomplete in the initial description. Will also take some time this weekend to make sure the max_fee argument gets passed through

File: cmd/lncli/commands.go [hunk 6]
```diff
@@ -84,6 +84,37 @@ func actionDecorator(f func(*cli.Context) error) func(*cli.Context) error {
...
+func calculateFeeLimit(ctx *cli.Context, amount int64) int64 {
+	var (
+		feeLimit        int64
+		feeLimitPercent int64
+	)
+
+	const MaxInt64Value int64 = 1<<63 - 1 // 9223372036854775807
```
Comment (halseth) [inline cmd/lncli/commands.go]: this constant exists in `math`

File: cmd/lncli/commands.go [hunk 7]
```diff
@@ -84,6 +84,37 @@ func actionDecorator(f func(*cli.Context) error) func(*cli.Context) error {
...
+		feeLimit = feeLimitPercent
+	}
+
+	if ctx.IsSet("fee_limit") {
+		feeLimit = ctx.Int64("fee_limit")
+	}
+
+	// If both fee_limit and fee_limit_percent are set, pick the minimum
+	if ctx.IsSet("fee_limit") && ctx.IsSet("fee_limit_percent") {
+		feeLimit = int64(math.Min(float64(feeLimit), float64(feeLimitPercent)))
```
Comment (halseth) [inline cmd/lncli/commands.go]: Can just take the minimum as you go

File: cmd/lncli/commands.go (hunk 1 above)
Comment (halseth) [reply cmd/lncli/commands.go]: Since it is entirely possible to use the `pay_req` without specifying the amount, I think the fee limit calculating probably should be done in the RPC server, after the pay req is parsed.

**Comment by halseth** [comment]
replaced by #1113.
